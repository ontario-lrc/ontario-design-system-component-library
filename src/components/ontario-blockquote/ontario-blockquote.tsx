import { Component, h, Element, Prop, State, Watch } from '@stencil/core';

import { Blockquote } from './blockquote.interface';

import { validatePropExists } from '../../utils/validation/validation-functions';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';

@Component({
	tag: 'ontario-blockquote',
	styleUrl: 'ontario-blockquote.scss',
	shadow: true,
})
export class OntarioBlockquote implements Blockquote {
	@Element() host: HTMLElement;

	/**
	 * Text to be displayed as the quote.
	 *
	 * Note that wrapping the quotes in quotations is not needed - this is handled through the component styles
	 */
	@Prop({ mutable: true }) quote: string;

	/**
	 * Optional text to be displayed as the attribution (the author) of the quote.
	 */
	@Prop() attribution?: string;

	/**
	 * Optional text to be displayed for additional information about the attribution/author.
	 */
	@Prop() byline?: string;

	@State() shortQuoteLength: number = 140;

	@State() shortQuote: boolean = false;

	@State() private quoteState: string;

	/*
	 * Watch for changes in the `quote` prop for validation purposes.
	 *
	 * If the `quote` prop is not provided, set the `quote` prop to the host textContent (if it exists).
	 * If the `quote` prop length is 140 characters or less, set the `shortQuote` state to true - this will add additonal classes for the blockquote styles.
	 */
	@Watch('quote')
	validateQuote() {
		this.quoteState = this.quote ?? this.host.textContent ?? '';
		this.validateQuoteContent(this.quoteState);
		this.shortQuote = this.quoteState?.length <= this.shortQuoteLength ?? true;
	}

	/**
	 * Print the missing `quote` prop warning message
	 */
	validateQuoteContent(newValue: string) {
		if (validatePropExists(newValue)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' quote ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-blockquote> ')
				.addRegularText('was not provided')
				.printMessage();
		}
	}

	componentWillLoad() {
		this.validateQuote();
	}

	/**
	 * This helper is used to help load translations for any slots + text content passed in by the user.
	 */
	componentDidLoad() {
		const observer = new MutationObserver((mutations) => {
			mutations.forEach((mutation) => {
				if (mutation.type === 'attributes') {
					this.validateQuote();
				}
			});
		});

		const options = { attributes: true };
		observer.observe(this.host, options);
	}

	render() {
		return (
			<blockquote class={this.shortQuote ? `ontario-blockquote ontario-blockquote--short` : `ontario-blockquote`}>
				<p>{this.quoteState}</p>
				{this.attribution && <cite class="ontario-blockquote__attribution">{this.attribution}</cite>}
				{this.byline && <cite class="ontario-blockquote__byline">{this.byline}</cite>}
			</blockquote>
		);
	}
}
