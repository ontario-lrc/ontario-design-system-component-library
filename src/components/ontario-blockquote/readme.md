# ontario-blockquote

Use a blockquote to draw attention to a speaker quote or excerpt.

## Usage guidance

Please refer to the [Ontario Design System](https://designsystem.ontario.ca/components/detail/blockquote.html) for current documentation guidance for blockquotes.

## Configuration

Once the component package has been installed (see Ontario Design System Component Library for installation instructions), the blockquote component can be added directly into the project's code, and can be customized by updating the properties outlined [here](#properties). Please see the [examples](#examples) below for how to configure the component.

## Examples

Example of a blockquote component where the user is explicitly passing in a quote through the `quote` property - this will override the `Quote Content` value, and the resulting quote text passed into the `quote` property will be displayed.

```html
<ontario-blockquote
	attribution="Survey respondent"
	quote="Access to high-quality child care is an issue that impacts our entire society."
>
	Quote Content
</ontario-blockquote>
```

This is another example of a long blockquote. The component calculates the length of the quote passed in through either the `quote` property or the blockquote content, and applies the appropriate classes and styles to display either a short or long blockquote appearance.

```html
<ontario-blockquote attribution="Academic partners" byline="Ontario's Pedagogy for the Early Years">
	In the past few years, Ontario has gained a high level of respect for its visionary work in early education. The
	changes have been profound, and thoughtfully introduced. The views that guide the work are articulated clearly, and
	express great respect for children, families, and educators. We now have a solid foundation upon which to build a more
	coherent system.
</ontario-blockquote>
```

<!-- Auto Generated Below -->

## Properties

| Property      | Attribute     | Description                                                                                                                                 | Type                  | Default     |
| ------------- | ------------- | ------------------------------------------------------------------------------------------------------------------------------------------- | --------------------- | ----------- |
| `attribution` | `attribution` | Optional text to be displayed as the attribution (the author) of the quote.                                                                 | `string \| undefined` | `undefined` |
| `byline`      | `byline`      | Optional text to be displayed for additional information about the attribution/author.                                                      | `string \| undefined` | `undefined` |
| `quote`       | `quote`       | Text to be displayed as the quote. Note that wrapping the quotes in quotations is not needed - this is handled through the component styles | `string`              | `undefined` |

---

_Built with [StencilJS](https://stenciljs.com/)_
