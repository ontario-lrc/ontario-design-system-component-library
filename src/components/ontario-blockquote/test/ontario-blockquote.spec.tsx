import { newSpecPage } from '@stencil/core/testing';
import { OntarioBlockquote } from '../ontario-blockquote';
import { mutationObserverMock } from '../../../utils/tests/mutation-observer.mock';

global.MutationObserver = mutationObserverMock;

describe('ontario-blockquote', () => {
	describe('snapshot', () => {
		it('should render the expected html', async () => {
			const page = await newSpecPage({
				components: [OntarioBlockquote],
				html: `<ontario-blockquote quote="This is a test quote" attribution="DS Tech Team"></ontario-blockquote>`,
			});

			expect(page.root).toMatchSnapshot();
		});
	});

	it('should render a byline when the byline prop is passed', async () => {
		const page = await newSpecPage({
			components: [OntarioBlockquote],
			html: `<ontario-blockquote attribution="DS Tech Team" byline="The best team at the ODS!">This is a test quote</ontario-blockquote>`,
		});
		expect(page.root).toEqualHtml(`
      <ontario-blockquote attribution="DS Tech Team" byline="The best team at the ODS!">
        <mock:shadow-root>
          <blockquote class="ontario-blockquote ontario-blockquote--short">
						<p>This is a test quote</p>
						<cite class="ontario-blockquote__attribution">DS Tech Team</cite>
						<cite class="ontario-blockquote__byline">The best team at the ODS!</cite>
					</blockquote>
        </mock:shadow-root>
				This is a test quote
      </ontario-blockquote>
    `);
	});

	it('should reflect attributes/props being set', async () => {
		const page = await newSpecPage({
			components: [OntarioBlockquote],
			html: `<ontario-blockquote quote="This is another test quote" attribution="DS Tech Team" byline="Ontario Digital Service"></ontario-blockquote>`,
		});

		expect(page.rootInstance.quote).toBe('This is another test quote');
		expect(page.rootInstance.attribution).toBe('DS Tech Team');
		expect(page.rootInstance.byline).toBe('Ontario Digital Service');
	});

	it('should set the shortQuote state to `true` when the character count is under 140', async () => {
		const page = await newSpecPage({
			components: [OntarioBlockquote],
			html: `<ontario-blockquote quote="This is another test quote" attribution="DS Tech Team" byline="Ontario Digital Service"></ontario-blockquote>`,
		});

		expect(page.rootInstance.shortQuote).toBe(true);
	});

	it('should set the shortQuote state to `false` when the character count is over 140', async () => {
		const page = await newSpecPage({
			components: [OntarioBlockquote],
			html: `<ontario-blockquote attribution="Homer Simpson">I saw this movie about a bus that had to speed around a city, keeping its speed over fifty, and if its speed dropped, it would explode! I think it was called, "The Bus That Couldnâ€™t Slow Down".</ontario-blockquote>`,
		});

		expect(page.rootInstance.shortQuote).toBe(false);
	});
});
