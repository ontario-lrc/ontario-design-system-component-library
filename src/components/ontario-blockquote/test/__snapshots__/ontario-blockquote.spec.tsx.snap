// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ontario-blockquote snapshot should render the expected html 1`] = `
<ontario-blockquote attribution="DS Tech Team" quote="This is a test quote">
  <mock:shadow-root>
    <blockquote class="ontario-blockquote ontario-blockquote--short">
      <p>
        This is a test quote
      </p>
      <cite class="ontario-blockquote__attribution">
        DS Tech Team
      </cite>
    </blockquote>
  </mock:shadow-root>
</ontario-blockquote>
`;
